File: python/app/pfm/sketch_waves.py
"""Sketch Waves PFM — A premium drawing algorithm for DrawingBotV3.

Algorithm:
  Draws sine wave lines that vary in amplitude based on image brightness.
"""

from __future__ import annotations
import math
import numpy as np
from typing import List, Callable

from app.pfm import (
    PathFindingModule, PFMSetting, SettingType, DrawingGeometry,
    make_plotting_resolution_setting, make_random_seed_setting,
    make_sketch_style_settings, make_sketch_erase_settings,
)
from app.core.geometry import Path


class SketchWavesPFM(PathFindingModule):
    @property
    def name(self) -> str: return "Sketch Waves"
    @property
    def description(self) -> str: return "Draws sine waves varying amplitude by brightness."
    @property
    def category(self) -> str: return "Sketch"
    @property
    def is_premium(self) -> bool: return True

    def _define_settings(self) -> List[PFMSetting]:
        settings = [
            make_plotting_resolution_setting(),
            make_random_seed_setting(),
            PFMSetting("lines_count", "Lines Count", SettingType.INTEGER, 100, 10, 1000, 10),
            PFMSetting("wave_freq", "Frequency", SettingType.NUMBER, 10.0, 1.0, 100.0, 1.0),
            PFMSetting("wave_amp", "Max Amplitude", SettingType.NUMBER, 20.0, 1.0, 100.0, 1.0),
        ]
        settings.extend(make_sketch_erase_settings())
        return settings

    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        lines_count = self.get("lines_count")
        freq = self.get("wave_freq")
        max_amp = self.get("wave_amp")
        
        geometries: List[DrawingGeometry] = []
        lightened = image.astype(np.float32)
        
        for i in range(lines_count):
            if self.is_cancelled: break
            
            y_base = (i / max(1, lines_count - 1)) * h
            path: Path = []
            
            for x in range(w):
                if y_base >= h: continue
                # Get brightness
                b = lightened[int(min(h-1, y_base)), int(x)] / 255.0
                # Darker = higher amplitude
                amp = max_amp * (1.0 - b)
                
                y = y_base + math.sin(x / w * math.pi * 2 * freq) * amp
                path.append((float(x), float(y)))
                
            geometries.append(DrawingGeometry(path=path))
            progress((i + 1) / lines_count, len(geometries), f"Drawing Waves {i+1}/{lines_count}")
            
        return geometries

File: cpp_backup/pfm/sketch_waves.cpp
#include "pfm/sketch_waves.h"
#include <cmath>

SketchWavesPFM::SketchWavesPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> SketchWavesPFM::defineSettings() const {
    return {
        { "plotting_resolution", "Plotting Resolution", SettingType::Number,  1.0,  {}, 0.05, 2.0,  0.1, 1.0,  0.05, {}, "Default", "Scale factor applied to the image before processing." },
        { "random_seed",         "Random Seed",         SettingType::Integer, 42,   {}, 0,    999999,0,   999999,1,   {}, "Default" },
        { "lines_count",         "Lines Count",         SettingType::Integer, 100,  {}, 10,   1000,  10,  500,   10 },
        { "wave_freq",           "Frequency",           SettingType::Number,  10.0, {}, 1.0,  100.0, 1.0, 50.0,  1.0 },
        { "wave_amp",            "Max Amplitude",       SettingType::Number,  20.0, {}, 1.0,  100.0, 1.0, 50.0,  1.0 },
    };
}

QVector<DrawingGeometry> SketchWavesPFM::_process(const cv::Mat& image) {
    const int h = image.rows, w = image.cols;
    const int   linesCount = get("lines_count").toInt();
    const double freq      = get("wave_freq").toDouble();
    const double maxAmp    = get("wave_amp").toDouble();

    QVector<DrawingGeometry> geometries;
    geometries.reserve(linesCount);

    const float* imgData = reinterpret_cast<const float*>(image.data);
    // Convert once to float for fast access
    cv::Mat imgF;
    image.convertTo(imgF, CV_32F);

    for (int i = 0; i < linesCount && !isCancelled(); ++i) {
        double yBase = (double)i / std::max(1, linesCount - 1) * h;
        Path path;
        path.reserve(w);

        for (int x = 0; x < w; ++x) {
            int yi = (int)std::min((double)(h - 1), yBase);
            float brightness = imgF.at<float>(yi, x) / 255.0f;
            double amp = maxAmp * (1.0 - brightness);
            double y = yBase + std::sin((double)x / w * M_PI * 2.0 * freq) * amp;
            path.push_back({ (double)x, y });
        }

        geometries.append(DrawingGeometry{ path, 0 });
        if ((i + 1) % 10 == 0)
            emitProgress((float)(i + 1) / linesCount, geometries.size(),
                         QString("Waves %1/%2").arg(i + 1).arg(linesCount));
    }
    return geometries;
}
